File: ui/js/mock/transport.js
/**
 * mock/transport.js — the simulated clock: play/pause, auto-advance on
 * getState polls, deterministic advanceBy stepping for tests, and the
 * published masterPos view (wrapped on the effective cycle at rest,
 * linear from the frozen base while recording).
 */

// --- Transport Simulation ---
// Simulates the C++ audio engine advancing masterPos and recording clip duration.
// Auto-advance mode hooks into getState() polls (a fixed sample step).
// Deterministic mode uses advanceBy() for exact sample-count stepping.

import { posMod } from '../math_utils.js';
import { state, effectiveQuantumForState, shiftOrigins, rootFrameTop } from './state.js';
import { effectiveCycle } from './cycles.js';
import { recView, growRecordingClips } from './recording.js';

/**
 * The simulation STEP: how far the clock moves per getState() poll.
 * Deliberately a fixed SAMPLE COUNT, NOT derived from the mock's sample
 * rate (mock/rate.js) — even though 2205 was originally chosen as
 * ~50 ms at 44.1 kHz.
 *
 * Why not derive it: a test that plays and then polls reads the clock
 * at multiples of this step, so a rate-derived step would move those
 * positions under a rate sweep and flip outcomes that fold on them
 * (map_anchor's cut then lands in a different Q). A mock tick is a
 * simulation step, not a wall-clock duration — keeping it in samples is
 * what lets the rest of the rate generalization sweep cleanly.
 */
export const DEFAULT_SAMPLES_PER_TICK = 2205;

/**
 * The UI polls every ~50 ms, so a fixed step means the mock renders a
 * CONSTANT amount of simulated audio per wall-clock second, whatever
 * the device rate is. Real-time specs (the dead-reckoning playhead
 * sweep) need this to know how long a loop takes on the wall clock:
 * `loopSeconds = loopSamples / SIMULATED_SAMPLES_PER_SECOND`.
 */
const POLL_INTERVAL_MS = 50;
export const SIMULATED_SAMPLES_PER_SECOND =
    DEFAULT_SAMPLES_PER_TICK * (1000 / POLL_INTERVAL_MS);

export const transport = {
    running: false,                            // Is transport auto-advancing?
    samplesPerTick: DEFAULT_SAMPLES_PER_TICK,  // simulation step, in samples
    speed: 1.0                                 // Speed multiplier (1.0 = real-time)
};

// Mirrors AudioEngine::togglePlayback — pause/resume: the clock is
// never reset (kernel.md); stopping freezes the view where it is.
export function togglePlayback() {
    state.isPlaying = !state.isPlaying;
    transport.running = state.isPlaying;
    console.log('[MockBackend] togglePlayback →', state.isPlaying);
    return true;
}

// Called on every getState() poll when transport is running
export function advanceTransport() {
    if (!transport.running) return;

    const advance = Math.round(transport.samplesPerTick * transport.speed);
    state.masterPos = state.masterPos + advance;
    growRecordingClips(state.nodes, advance);
}

// Start auto-advancing transport (hooks into getState polls)
export function startTransport(speed = 1.0) {
    transport.running = true;
    transport.speed = speed;
    state.isPlaying = true;
    console.log(`[MockBackend] Transport started (speed=${speed})`);
}

// Pause auto-advancing transport
export function pauseTransport() {
    transport.running = false;
    console.log('[MockBackend] Transport paused');
}

/**
 * Ruler scrub (engine parity: AudioEngine::seekTransport). THE SEEK IS
 * A PHASE ADVANCE (docs/frame.md): the backend reads no frame — the
 * view seats the frame's zero — so the view sends how far the playing
 * phase should move, computed against the raw clock it last polled
 * (`atClock`); the backend corrects for the clock having moved since.
 * The mock's clock is monotonic like the engine's (kernel.md), so
 * advancing the phase moves the island's zero — and every origin with
 * it — BACK by the advance, never masterPos. Refused (false) while any
 * take is live or armed — takes place audio by the clock. NOT undoable
 * (a monitoring gesture — keep it out of mock/undo.js's intercept set).
 */
export function seekTransport(deltaSamples, atClock) {
    if (recView.active || anyClipHot(state.nodes)) return false;
    let advance = Math.round(Number(deltaSamples) || 0);
    if (Number.isFinite(atClock)) advance -= state.masterPos - Math.round(atClock);
    // Answers what was applied and when (engine parity), so the view
    // keeps its frame facts exact between polls.
    const applied = { advance, clock: state.masterPos };
    const delta = -advance;
    if (delta === 0) return applied;
    state.islandZero = (state.islandZero || 0) + delta;
    // A seek is a phase jump of the whole island (composition.md §5,
    // engine parity AudioEngine::seekTransport): every origin — clips
    // AND stacks (Q18), the root's too — rides the zero's delta, so
    // placement on the grid (origin − zero) is unchanged and playback
    // lands at the requested phase. shiftOrigins(root, delta) is the
    // one primitive.
    state.nodes.forEach(n => shiftOrigins(n, delta));
    if (state.rootAnchored) state.rootOrigin += delta;
    console.log(`[MockBackend] seekTransport → phase +${advance} (zero=${state.islandZero})`);
    return applied;
}

/** Any clip live or armed anywhere in the graph (the engine's
 * hasActiveTake ∨ isArmedOrRecording twin). */
function anyClipHot(nodes) {
    return (nodes || []).some(n =>
        n.isRecording || n.isPendingStart || anyClipHot(n.nodes));
}

// Deterministic advance by exact sample count (for reliable test assertions)
export function advanceBy(samples) {
    state.masterPos = state.masterPos + samples;
    growRecordingClips(state.nodes, samples);
    console.log(`[MockBackend] Advanced by ${samples} samples → masterPos=${state.masterPos}`);
}

/**
 * masterPos CONTRACT (mirrors AudioEngine::getGraphState — kernel.md
 * step 3): the engine's clock is monotonic and never exposed raw. The
 * published masterPos is a DERIVED VIEW — wrapped to the cycle when
 * idle/playing, but during recording it grows linearly from a base
 * frozen at record start, so the cursor extends past the committed LCM.
 * Consumers must NOT re-wrap it (re-wrapping makes a growing take loop
 * 1Q over and over).
 */
export function viewMasterPos() {
    const raw = state.masterPos;
    if (recView.active) return recView.base + (raw - recView.anchor);
    const Q = effectiveQuantumForState();
    // E-C (engine parity): the view wraps on the EFFECTIVE cycle — the
    // playhead loops with what is heard, never past an active window.
    const cycle = effectiveCycle(Q);
    const rel = raw - rootFrameTop(); // engine: rel = t − rootFrameTop()
    return cycle > 0 ? posMod(rel, cycle) : rel;
}

File: src/engine/transport.cc
// AudioEngine — TRANSPORT + STATE: play/pause, the seek (a whole-island
// phase jump whose delta the undo/redo logs ride), getGraphState — the
// ONE UI state publication — with its transport block, waveform reads
// and the effective-cycle helper. Message thread only.

#include "../audio_engine.h"

#include <cmath>

#include "../clip_node.h"
#include "../rt_log.h"
#include "../stack_node.h"
#include "../timing.h"
#include "engine_internal.h"


void AudioEngine::togglePlayback() {
  // Pause/resume: stopping freezes the clock where it is; playing
  // resumes from the same phase. The clock is never reset (kernel.md).
  // (Restart-from-top is not an engine concept; if it ever were, it
  // would be a root time-map — kernel.md §3.) The user-facing play
  // start is UI policy composed from this toggle and seekTransport —
  // session_view.md display law 15, ui/js/play_start.js. No pause under a live take
  // (owner ruling 2026-09-09): the take performs against the running
  // clock — finish or cancel it first.
  if (is_playing_global.load() && refusedUnderLiveTake("pause")) return;
  is_playing_global = !is_playing_global.load();
}

bool AudioEngine::seekTransport(double delta_samples,
                                std::optional<int64_t> at_clock,
                                SeekResult* applied) {
  // Refused while any take is live or armed: takes place audio by
  // this clock (arm targets, origins, commit boundaries all read it),
  // so a mid-take phase jump would corrupt the take's placement. The
  // UI mirrors the refusal (ruler shows a locked cursor), but the
  // engine owns the rule.
  if (root_node == nullptr) return false;
  // Settle any take the audio thread committed since the last poll
  // FIRST: shiftHistoryAbsolutes below re-frames the log, and a take
  // that has not entered it yet would carry a pre-seek zero into its
  // later Untake entry (record/undo/redo reconcile the same way).
  reconcileTakes();
  if (root_node->hasActiveTake() || root_node->isArmedOrRecording()) {
    return false;
  }

  // THE SEEK IS A PHASE ADVANCE (docs/frame.md): the engine reads no
  // frame — only the view knows where the frame's zero is seated — so
  // the view sends how far the playing phase should move, computed
  // against the transport reading it last polled (`at_clock`). The
  // clock has moved since that poll; correcting for it lands the phase
  // the view meant, exactly, stopped or playing. Advancing the phase
  // by `advance` is moving the island's zero — and every origin with
  // it — BACK by `advance`. The monotonic clock is untouched
  // (kernel.md); islandPos teleports with the zero, and the UI's
  // dead-reckoner classifies the jump as a TELEPORT, never velocity
  // (playhead_clock.js). A zero advance is an accepted no-op.
  const int64_t t = global_transport_pos.load();
  int64_t advance = (int64_t)std::llround(delta_samples);
  if (at_clock.has_value()) advance -= t - *at_clock;
  if (applied != nullptr) {
    applied->advance = advance;
    applied->clock = t;
  }
  const int64_t delta = -advance;
  if (delta == 0) return true;
  const int64_t zero_old = root_node->getZero();
  const int64_t zero_new = zero_old + delta;
  const uint32_t gen = root_node->nextIslandGeneration();
  // THE CONTENT-FRAME LAW (time_maps.md; pinned by
  // tests/content_frame_tests.cc): clips read their buffers
  // ORIGIN-relative on the monotonic clock, so moving the zero alone
  // would move the cursor and NOT the audio. A seek is a phase jump of
  // the whole island: every origin rides the zero's delta, so each
  // clip's placement on the grid (origin − zero) is unchanged and
  // playback lands at the requested phase. Not undoable, like the seek
  // itself.
  // Every origin — clips AND stacks (Q18), the root's too — rides the
  // delta: the recursive shift is the one primitive (composition.md §5).
  shiftOriginsGated(*root_node, delta, gen);
  // Origins first (gated), then the zero with the generation: one
  // block top adopts both or neither.
  root_node->seekZeroTo(zero_new, gen);
  // THE HISTORY RIDES TOO: the undo/redo logs store ABSOLUTE origins
  // and zeros, and a seek re-frames every absolute in the session. An
  // inverse restoring pre-seek absolutes for a SUBSET of clips (a
  // continuity rider, a take payload) would shift that subset against
  // everything else — undo would audibly move a clip the edit never
  // touched. Shift every absolute in both logs by the same delta, so
  // undo after a seek restores the same PLACEMENT.
  if (delta != 0) shiftHistoryAbsolutes(delta);
  return true;
}

void AudioEngine::shiftHistoryAbsolutes(int64_t delta) {
  const std::function<void(celestrian::AudioNode*)> shiftSubtree =
      [&](celestrian::AudioNode* node) {
        if (node == nullptr) return;
        // Detached subtrees held by the log: every node's origin (Q18 —
        // stacks too), ungated (nothing renders them).
        node->origin_samples.store(node->origin_samples.load() + delta);
        if (node->getNodeType() == celestrian::NodeType::Clip) return;
        auto* stack = static_cast<celestrian::StackNode*>(node);
        for (const auto& child : stack->ownedChildren())
          shiftSubtree(child.get());
      };
  auto shiftEdit = [&](celestrian::Edit& e) {
    // setsIsland's izero is a real zero on every kind that sets it.
    // A Collapse inverse's facts (shift, old_duration, the window) are
    // RELATIVE; its splice form's pre-splice origin rides `iorg` under
    // setsOrigin and shifts below like every absolute.
    if (e.setsIsland) e.izero += delta;
    if (e.setsOrigin) e.iorg += delta;
    for (auto& r : e.anchors) r.origin += delta;
    for (auto& r : e.seq_riders) r.origin += delta;  // the root's anchor
    for (auto& tp : e.takes) tp.state.origin += delta;
    for (auto& ot : e.other_takes) ot.second.origin += delta;
    shiftSubtree(e.node.get());
    shiftSubtree(e.node2.get());
  };
  for (auto& e : undo_) shiftEdit(e);
  for (auto& e : redo_) shiftEdit(e);
}

void AudioEngine::tick() {
  // Forward any log lines queued by the audio thread.
  celestrian::RtLog::instance().drain();
  // Settled takes enter the undo log (see PendingTake), and every live
  // take keeps its storage headroom.
  reconcileTakes();
  growLiveTakes();
}

juce::var AudioEngine::getGraphState() {
  tick();

  // Cycle view of the monotonic clock (kernel.md step 3): the engine
  // never wraps its transport; the UI-facing masterPos is derived here.
  // Idle/playback: t mod LCM. Recording: frozen base + linear growth so
  // the cursor extends past the committed LCM.
  const int64_t t = global_transport_pos.load();
  double master_view;
  if (view_recording_.load()) {
    master_view = (double)(view_base_.load() + (t - view_anchor_t_.load()));
  } else {
    // Wrap on the EFFECTIVE cycle (E-C): active windows shorten what
    // is audible, and the playhead loops with what is heard — from the
    // root's frame top (its song's origin while it carries one, else
    // the island zero), the place the root's own geometry folds from.
    const int64_t cycle = calculateEffectiveCycleLength();
    const int64_t rel = t - rootFrameTop();
    master_view = (double)celestrian::timing::posMod(rel, cycle);
  }
  // The RAW island clock (zero-relative, unwrapped): masterPos above is
  // folded on the CURRENT audible cycle, so its fold point jumps when a
  // live map edit changes that cycle mid-gesture. The UI folds this
  // invariant clock on its own (pinned) frame for a continuous cursor.
  const double island_view = (double)(t - islandZero());

  if (root_node) {
    auto metadata = root_node->getMetadata();
    auto* obj = metadata.getDynamicObject();
    attachTransportState(*obj, master_view, island_view);
    // NOTE: the root's metadata already carries `quantum` (its stored
    // island Q, stack_node.cc) — the VM reads it as the top-level Q
    // fact instead of re-deriving min-over-nodes.
    obj->setProperty("focusedId", root_node->getUuid());
    return metadata;
  }

  juce::DynamicObject::Ptr state = new juce::DynamicObject();
  attachTransportState(*state, master_view, island_view);
  state->setProperty("quantum",
                     (double)(root_node ? root_node->getQuantum() : 0));
  state->setProperty("nodes", juce::Array<juce::var>());
  return juce::var(state.get());
}

void AudioEngine::attachTransportState(juce::DynamicObject& state,
                                       double master_view,
                                       double island_view) const {
  state.setProperty("isPlaying", (bool)is_playing_global.load());
  state.setProperty("masterPos", master_view);
  state.setProperty("islandPos", island_view);
  // The island zero is the UI's frame origin for every cycle-relative
  // projection (kernel.md one-frame rule). It is NOT the root node's
  // `origin` metadata — commit re-bases the zero (StackNode::takeCommitted),
  // and the UI marking take-vs-ghost tiles needs the re-based value.
  state.setProperty("islandZero", (double)islandZero());
  // THE DEFINER, published: the sole committed
  // clip, or the definer stack, whose window re-establishes Q — the UI
  // reads this instead of re-deriving it with its own (drifting)
  // definition. Empty when the island has no definer.
  {
    // engine_internal::definer carries EVERY gate (sole geometry, no
    // live take, no audition) — the UI reads the answer, never re-adds
    // a gate.
    auto* d = celestrian::engine_internal::definer(*root_node);
    state.setProperty("definerId", d != nullptr ? d->getUuid() : juce::String());
  }
  // Master monitor: smoothed output RMS per channel (linear 0..1) —
  // the transport VU section reads these off the state poll.
  state.setProperty("masterVuL", (double)master_vu_l_.load());
  state.setProperty("masterVuR", (double)master_vu_r_.load());
  state.setProperty("canUndo", canUndo());
  state.setProperty("canRedo", canRedo());
  state.setProperty("perf", makePerfState());
}

juce::var AudioEngine::getWaveform(const juce::String& uuid,
                                   int num_peaks) const {
  auto* self = const_cast<AudioEngine*>(this);
  if (auto* node = self->findNodeByUuid(root_node.get(), uuid)) {
    return node->getWaveform(num_peaks);
  }
  return juce::Array<juce::var>();
}

juce::var AudioEngine::getTakeWaveform(const juce::String& uuid, int index,
                                       int num_peaks) const {
  auto* self = const_cast<AudioEngine*>(this);
  if (auto* clip = dynamic_cast<celestrian::ClipNode*>(
          self->findNodeByUuid(root_node.get(), uuid))) {
    return clip->getTakeWaveform(index, num_peaks);
  }
  return juce::Array<juce::var>();
}

// --- LCM Timeline Helpers ---

int64_t AudioEngine::calculateEffectiveCycleLength() const {
  // THE PERIOD LAW over the current snapshot (period_law.h /
  // graph_snapshot.h) — the very computation the audio callback seeds
  // its context with, so the message thread and the callback cannot
  // disagree on the cycle. The snapshot is immutable once published and
  // the message thread publishes it; node facts (durations, windows,
  // sequences) are atomics read live.
  const int64_t one_second = (int64_t)cached_sample_rate_.load();
  if (!root_node) return one_second;
  const auto* snap = graph_snapshot_.load(std::memory_order_acquire);
  if (snap == nullptr) return one_second;
  return celestrian::snapEffectiveCycle(*snap, root_node->getEffectiveQuantum(),
                                        one_second);
}
